staging/src/k8s.io/dynamic-resource-allocation/controller/controller.go

```diff
@@ -85,7 +86,7 @@ type Driver interface {
 	//
 	// The objects are read-only and must not be modified. This call
 	// must be idempotent.
-	Allocate(ctx context.Context, claim *resourcev1alpha2.ResourceClaim, claimParameters interface{}, class *resourcev1alpha2.ResourceClass, classParameters interface{}, selectedNode string) (*resourcev1alpha2.AllocationResult, error)
+	Allocate(ctx context.Context, claims []*ClaimAllocation, selectedNode string) ([]*resourcev1alpha2.AllocationResult, error)
```

Comment:
What is supposed to happen when only some claims can be allocated? This needs to be documented and has an impact on the implementation, both in the generic controller and in drivers.

I haven't looked at what is being done now, but I suspect the failure in ["multiple nodes reallocation works"](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/118862/pull-kubernetes-kind-dra/1674037982060875776) comes from changing the behavior compared to what it was before.

Previously, every allocated claim was updated to record that it was allocated.

Reply:
I did not change the logic that picks which claims can be allocated, previously the allocation for claims was called in the loop, I am simply replacing the loop with one call accepting a slice of claims as an argument, moving the looping through claims into the driver's responsibility.

Reply:
That's a good point, I did not think about partial failure, only all or nothing. Why can we not assume that all drivers can allocate all claims atomically, is there a number of attempts we give driver to allocate a claim before another node has to be chosen for the Pod, is it not instant - after failure to allocate particular claim all currently allocated claims (given they are only booked for single pod) are deallocated and pod is moved to another node?

Returning less items in `[]*resourcev1alpha2.AllocationResult` than incoming claims should be treated as a driver's failure to respond properly, AllocationResult has no reference inside of to which claim it is meant to be added to, so the only allowed correct allocation response from driver is same amount and precisely the same order of `[]*resourcev1alpha2.AllocationResult` items (objects or `nil`s) that the incoming claims are in.

Would it make sense to leave `err` for something fatal like ResourceClass for claims is not found, or Pod not found - something common for all claims, and in case `err` is set - disregard the whole slice of allocationResults, but if nothing dramatic happened, treat the list of `AllocationResult`s as a success/failure based on whether the particular element is nil or not?

Reply:
> Why can we not assume that all drivers can allocate all claims atomically

Because the driver might manage network-attached resources where each claim is independent from the other and the underlying backend APIs don't support batching. Then the driver has to implement this API by iterating over claims and each individual allocation may succeed or fail.

When one claim fails, it doesn't make sense to deallocate because it implies always starting from scratch, with a lot of churn, instead of just trying again for the remaining unallocated claims. The current approach in kube-scheduler is to keep claims allocated and unreserved and only trigger deallocation if needed because a new node had to be chosen.

> Returning less items in []*resourcev1alpha2.AllocationResult than incoming claims should be treated as a driver's failure to respond properly

Agreed. With that in mind, what if we designed this API so that a driver can't make that mistake?

We could add `Err error` and `Allocation *resourcev1alpha2.AllocationResult` to `ClaimAllocation`. Then `Allocate` doesn't need to allocate and a return a slice of a suitable size. Instead it just sets those two fields.

> Would it make sense to leave err for something fatal like ResourceClass for claims is not found, or Pod not found - something common for all claims

We can do that. The semantic of non-nil error return then is "ignore slice, nothing was allocated, this error applies to all claims".

> if nothing dramatic happened, treat the list of AllocationResults as a success/failure based on whether the particular element is nil or not?

That would remove the ability of getting more information from the driver about why a particular claim couldn't get allocated. For that we need a per-claim error.

Reply:
> Agreed. With that in mind, what if we designed this API so that a driver can't make that mistake?
> 
> We could add Err error and Allocation *resourcev1alpha2.AllocationResult to ClaimAllocation. Then Allocate doesn't need to allocate and a return a slice of a suitable size. Instead it just sets those two fields.

I'll give it a try. I agree, it would give more visibility for per-claim errors reporting.